### ISSUE TITLE
ImageDataGenerator: userwarning + weird numbers

### ISSUE THREAD

**Issue by FiReTiTi** [fchollet/keras#5193]
Hi,
I've created my own ImageDataGenerator.
It reads images from two directories (one per class), and if the batch size is N, then each batch will contain N/2 images of each class.
And if there are M images per class, so the sample per epoch must be samples_per_epoch=2*M/N.

I did a little test:
 - 5000 images per class
 - batch of size 32
 - 3 epochs
 => samples_per_epoch = 312.

But I get this warning:

> UserWarning: Epoch comprised more than `samples_per_epoch` samples, which might affect learning results. Set `samples_per_epoch` correctly to avoid this warning.

Why?

Moreover, 
But when I do the training, fit_generator displays:

> Epoch 2/3
> 320/312 [==============================] - 0s - loss: 0.7812  

Why do I get 320/312?

Thanks for your help.

**Comment by patyork** [comment]
312 / 32 = 9.75. Keras doesn't do a batch of that remainder .75 * batch_size (or .75*32 = 24), so it does a full 10th batch. 10 * 32 == 320.

Thus, the warning letting you know that it didn't do exactly 312 samples in that epoch, because it has to do an integer number of batches (in this case 10) that are all of batch_size (in this case 32).

**Comment by FiReTiTi** [comment]
Thanks for your answer, but I am the one generating the batches, it's my own ImageDataGenerator.
312 is supposed to be the number of batches 2x5000/32 = 312.5
According to what you say, ok for an extra batch (that I do not generate), but why 320, so 8 extra batches?

**Comment by patyork** [comment]
You've a fundamental misunderstanding of what a batch, sample, and epoch are.

- **sample**: one data point (in this case, one image)
- **batch**: a set of `batch_size` samples that are run in parallel to generate one update statement
- **epoch**: an arbitrary delineator to separate training into separate phases
- **samples_per_epoch**: an arbitrary number of samples to comprise an "epoch" (often set to be the total number of samples in the training data set).
  - an **epoch** will therefore consist of `ceil(samples_per_epoch / batch_size)` batches, all of size `batch_size`

You've set `batch_size = 32` and `samples_per_epoch=312`. This means that in each epoch, you wish to train on 312 examples, or 9.75 batches of size 32. Keras can't and won't do a "partial" batch, so 9.5 is rounded up to 10. 10 batches * 32 samples per batch = 320 total samples. `320 <> 312`, thus the warning message and the progress bar that says "320/312" or "actual samples seen/samples_per_epoch parameter"

From the sounds of it, you want to set `batch_size = 32` and `samples_per_epoch = 10000`. 10,000 is still not evenly divisible by 32, so Keras will actually run 313 batches, totaling 313*32=10,016, which will generate a warning and a progress bar that reads "10016/10000".

**Comment by FiReTiTi** [comment]
Thanks, I misunderstood the parameter `samples_per_epoch` with `batches_per_epoch`.
